coap: examples cleanup

aabadie opened the pull request:
This PR proposes what I suggested in #6038, see [comment](https://github.com/RIOT-OS/RIOT/issues/6038#issuecomment-329168395).

The READMEs of the tests applications have been removed and the gcoap example has slightly been updated. The `/riot/board` handler has been moved from nanocoap to gcoap. A PUT/POST is also possible on the `/cli/stats` handler.

kb2ma commented on examples/gcoap/README.md:
```diff
@@ -53,6 +67,18 @@ CLI output:
 
+## Other available CoAP implementations and applications
+
+RIOT is compatible with other CoAP implementations with associated test
```
"compatible" is kind of general. Might be clearer to say: 

RIOT also provides package imports and test applications for other CoAP implementations:

kb2ma commented on examples/gcoap/README.md:
```diff
@@ -53,6 +67,18 @@ CLI output:
+
+RIOT is compatible with other CoAP implementations with associated test
+applications:
+* [Nanocoap](../tests/pkg_nanocoap): a very lightweight CoAP server based on
```
Requires an empty line before the list to format correctly in ReText. Formats OK in GitHub, so your choice.

Also, Nanocoap -> Microcoap

kb2ma replied:
I just tested these links, and the Nanocoap and Microcoap links don't work on GitHub. Need to pop up one more directory level.

kb2ma commented on examples/gcoap/gcoap_cli.c:
```diff
@@ -32,11 +32,14 @@
 /* CoAP resources */
 static const coap_resource_t _resources[] = {
-    { "/cli/stats", COAP_GET, _stats_handler },
+    { "/cli/stats", COAP_GET | COAP_PUT | COAP_POST, _stats_handler },
```
The function simply replaces the resource value, so I think this entry should include only COAP_PUT.

aabadie replied:
You are right

kb2ma commented on examples/gcoap/gcoap_cli.c:
```diff
@@ -93,11 +96,38 @@ static void _resp_handler(unsigned req_state, coap_pkt_t* pdu,
+            gcoap_resp_init(pdu, buf, len, COAP_CODE_CHANGED);
+
+            /* convert the payload to an integer and update the request count value */
+            char payload[16] = { 0 };
```
Shouldn't the length be 6 here? Max value is 65535.

kb2ma commented on examples/gcoap/gcoap_cli.c:
```diff
@@ -93,11 +96,38 @@ static void _resp_handler(unsigned req_state, coap_pkt_t* pdu,
+
+            /* convert the payload to an integer and update the request count value */
+            char payload[16] = { 0 };
+            memcpy(payload, (char*)pdu->payload, pdu->payload_len);
```
Don't copy more than payload can handle.

kb2ma commented on examples/gcoap/gcoap_cli.c:
```diff
@@ -93,11 +96,38 @@ static void _resp_handler(unsigned req_state, coap_pkt_t* pdu,
+            /* convert the payload to an integer and update the request count value */
+            char payload[16] = { 0 };
+            memcpy(payload, (char*)pdu->payload, pdu->payload_len);
+            req_count = strtol(payload, NULL, 10);
```
Use strtoul() and cast the right-hand result to uint16_t.

aabadie replied:
In fact, the PUT handler was not working. I fixed it.

kb2ma commented on examples/gcoap/README.md:
```diff
@@ -53,6 +67,18 @@ CLI output:
+* [Nanocoap](../tests/pkg_nanocoap): a very lightweight CoAP server based on
+  [nanocoap library](https://github.com/kaspar030/sock/tree/master/nanocoap)
+  implementation
+* [Nanocoap](../tests/pkg_microcoap): another lightweight CoAP server based on
```
Nanocoap -> Microcoap

aabadie commented on examples/gcoap/README.md:
```diff
@@ -1,19 +1,29 @@
+# gcoap example
 
-This application provides command line access to gcoap, a high-level API for CoAP messaging. See the [CoAP spec][1] for background, and the Modules>Networking>CoAP topic in the source documentation for detailed usage instructions and implementation notes.
+This application provides command line access to gcoap, a high-level API for
```
I did that to keep the 80 characters line width.

aabadie commented:
> why did you remove the READMEs for nanocoap and microcoap?

I initially thought they were useless. But in fact, they explain how to use coap clients (libcoap, copper). I'll keep the nanocoap one, since the microcoap one is the same.

kb2ma commented on examples/gcoap/README.md:
```diff
@@ -53,6 +67,19 @@ CLI output:
+  [nanocoap library](https://github.com/kaspar030/sock/tree/master/nanocoap)
+  implementation
+
+* [Micropcoap](../tests/pkg_microcoap): another lightweight CoAP server based
```
Still a typo. ;-)

kb2ma commented on examples/gcoap/gcoap_cli.c:
```diff
@@ -93,11 +96,39 @@ static void _resp_handler(unsigned req_state, coap_pkt_t* pdu,
+        case COAP_PUT:
+        {
+            /* Read and convert the content of the payload */
+            req_count = (uint16_t)strtoul((char*)pdu->payload, NULL, 10);
```
strtoul() expects the first argument to be a null-terminated string. pdu->payload is not guaranteed to be null-terminated. This is why the PUT/POST handler from nanocoap example created the temporary buffer.

Below is my solution, which worked for me on native. I added the bad request response because a payload that is too long is nonsense, and it also illustrates how easy it is to send an error response. You could also say that any non-zero value is nonsense, but I didn't go that far. :-)

Also, you don't need to send a payload with the success response. If the response is a 2.04, the sender knows they succeeded.

```
        case COAP_PUT:
            /* convert the payload to an integer and update the internal value */
            if (pdu->payload_len <= 5) {
                char payload[6] = { 0 };
                memcpy(payload, (char*)pdu->payload, pdu->payload_len);

                req_count = (uint16_t)strtoul(payload, NULL, 10);

                return gcoap_response(pdu, buf, len, COAP_CODE_CHANGED);
            }
            else {
                return gcoap_response(pdu, buf, len, COAP_CODE_BAD_REQUEST);
            }

```

I would also update the comment for this handler function to make the handling clear.

```
/*
 * Server callback for /cli/stats. Accepts either a GET or a PUT.
 *
 * GET: Returns the count of packets sent by the CLI.
 * PUT: Updates the count of packets. Rejects an obviously bad request, but
 *      allows any two byte value for example purposes. Semantically, the only
 *      valid action is to set the value to 0.
 */

```

aabadie commented:
Thanks @kb2ma, I addressed your comments and modified the `riot/board` handler to make it also use gcoap API.

kb2ma commented on examples/gcoap/gcoap_cli.c:
```diff
@@ -88,15 +91,50 @@ static void _resp_handler(unsigned req_state, coap_pkt_t* pdu,
+{
+    gcoap_resp_init(pdu, buf, len, COAP_CODE_CONTENT);
+    /* write the response buffer with the RIOT board name */
+    size_t payload_len = snprintf((char *)pdu->payload,
```
Good catch on the nanocoap function. What is the advantage here of snprintf() over a memcpy() like below?

```
    memcpy(pdu->payload, RIOT_BOARD, strlen(RIOT_BOARD));
    return gcoap_finish(pdu, strlen(RIOT_BOARD), COAP_FORMAT_TEXT);

```

aabadie replied:
>  What is the advantage here of snprintf() over a memcpy() like below?

memory consumption otherwise both works the same. Since your proposition consumes less memory, let's keep it.

kb2ma commented on examples/gcoap/gcoap_cli.c:
```diff
@@ -88,16 +91,50 @@ static void _resp_handler(unsigned req_state, coap_pkt_t* pdu,
+    unsigned method_flag = coap_method2flag(coap_get_code_detail(pdu));
+
+    switch(method_flag) {
+        case COAP_GET:
```
Minor issue: cases should be at same indent level as switch.

aabadie replied:
well, looking at other places in the code, this seems to be the more common style.

kb2ma replied:
I'm trying to do better myself [1], and comply with the coding conventions. :-) I don't feel strongly about this issue. 

[1] https://github.com/RIOT-OS/RIOT/pull/7337/files#diff-1ad46e1aa545fa54a91a5563e03a4ee8

aabadie commented:
@kb2ma, because of discussion during the Summit this week, I changed a few things again in this PR:
* reverted the nanocoap move: it seems that a decision was taken to keep it in the examples...
* updated the nanocoap README
* kept the microcoap_server README with the test application

kb2ma commented:
ACK. I tested microcoap test, nanocoap example, and gcoap example on native. The update to the nanocoap README was helpful. Thanks for persevering, @aabadie!

aabadie commented:
And thanks for reviewing @kb2ma ;)

Anyone else to double check and give an ACK ? @kaspar030, @cgundogan ?

aabadie commented:
> Anyone else to double check and give an ACK ? @kaspar030, @cgundogan ?

ping

aabadie commented:
Would be great to have this one in the coming release.

aabadie commented:
Thanks :)
